### ISSUE TITLE
Improve Conntrack garbage collection

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#5048]
The current implementation (as of v1.2) for managing the contents of the datapath connection tracking map leaves something to be desired: Once per minute, the userspace `cilium-agent` makes a series of calls to the `bpf()` syscall to fetch all of the entries in the map to determine whether they should be deleted. For each entry in the map, 2-3 calls must be made: One to fetch the next key, one to fetch the value, and perhaps one to delete the entry. The maximum size of the map is 1 million entries, and if the current count approaches this size then the garbage collection goroutine may spend a significant number of CPU cycles iterating and deleting elements from the conntrack map.

We have previously attempted using an LRU map to manage entries, however we hit https://github.com/cilium/cilium/issues/3239 . Observation was that TCP CT entries would be removed from the table and this would cause traffic disruption for long-lived connections.

**EDIT 2019-08**: As of v1.6 (and likely earlier releases) we have resolved some of the LRU issues and are back to using LRU for the primary method of managing conntrack entries.

Proposals to address this:

1. ~Build a userspace shadow of CT entries based on monitor events (related: #4080)~

   This would reduce the 2-3 syscalls per entry down to 1 syscall for each entry that needs to be deleted.

   * Ensure that trace messages are generated on datapath add/delete from CT map
     * Extend messages if necessary to include explicit notification of CT create/delete
   * Open a monitor socket from `cilium-agent`:
     * Create an entry in userspace when CT add event is received
     * Update timer when event is fired for packet allowed to a destination
     * Delete an entry when CT delete event is received
   * Periodically iterate over the userspace map and delete entries that have exceeded their timer
     * Sort CT entries by oldest and only iterate through entries whose timer is in the past

1. Manage CT entries more actively from the datapath

   * [x] Consider splitting TCP CT map from UDP map (#5501)
     * [ ] Populate TCP map via socket events
     * [ ] Replace UDP map GC from userspace with LRU map in datapath (on supported kernels)
   * [ ] Remove userspace management of CT map

**Comment by joestringer** [comment]
Problem discussed on the proposal (1) above was that the ringbuffer for monitor events from the datapath can lose events. If we lose events, we would lose understanding of the contents of the CT table, so could end up leaving entries behind because userspace doesn't know about them. To address this, would still need a periodic sweeper.

I've been investigating (2), I have some local WIP code for splitting the tables.

Separately, @tgraf is cooking something different where dummy entries are inserted/deleted into the table prior to a dump which has promise for fixing the LRU issue.

**Comment by tgraf** [comment]
#5501 will be handled in the 1.3 release. The remaining work is moved to 1.4

**Comment by joestringer** [comment]
We could likely build upon the new bpf_sock support here to populate the CT table:
https://github.com/cilium/cilium/blob/master/bpf/bpf_sock.c

**Comment by aanm** [comment]
Once this is done, stop doing this
https://github.com/cilium/cilium/blob/e4827c2eb6da82f96d1652abdb5a9b9fe4df0beb/pkg/endpoint/bpf.go#L531-L551

**Comment by kenieevan** [comment]
hello, guys, beside the trace/monitor method , is it possible to impelment a kernel module to traverse the bpf map, and do the filter, delete. Using this way, system call is avoided, and you can take rcu lock most of the time.

**Comment by kenieevan** [comment]
@joestringer thanks for your reply.   i did some experiement and find that  compared with ipvs+iptables SNAT, bpf  cps  is  140%  better.   our main concern is that LRU maybe not secure enough.  when there are a lot of new connection coming,  some long and idle connections may be evicted even they are not timeout.   I think the best solution maybe add timer entry in the bpf hash map, however there is no such code in main trunk now.  As to the userspace GC, it is nice and very necessary for us to  remedy the security issue by putting a  up limit on the timeout entry. However, when the net traffic is very busy, the GC  tends to issue too many syscalls.  In such case, it seems worth the effort to move the GC into kernel module.

**Comment by joestringer** [comment]
Regarding the "too many syscalls" issue, I've been informed that as of Linux v5.6 there is batched map lookup which should allow us to make this more efficient on newer kernels:

https://lore.kernel.org/bpf/20191119014357.98465-3-brianvv@google.com/

**Comment by jeff-an** [comment]
Hey there, we are noticing slowness in endpoint creation and deletion on an enterprise-scale K8s cluster running Cilium and we suspect conntrack cleaning might be part of the cause. 

We empirically notice endpoint creation take much longer when the node has heavy CPU workloads running on it, especially when many pods are scheduled at once. Debug logs seem to point to ct cleaning to be the longest factor (waitingForCTClean = 1 minute 3 seconds):

```
level=warning msg="Regeneration of endpoint failed" bpfCompilation=0s bpfLoadProg="344.712µs" bpfWaitForELF="10.29µs" bpfWriteELF="774.213µs" containerID= datapathPolicyRevision=0 desiredPolicyRevision=18 endpointID=229 error="Failed to replace Qdisc for lxc69bc0bfc43ce: Link not found" identity=45981 ipv4= ipv6= k8sPodName=/ mapSync="110.719µs" policyCalculation="76.11µs" prepareBuild="864.166µs" proxyConfiguration="20.04µs" proxyPolicyCalculation="32.43µs" proxyWaitForAck=0s reason="updated security labels" subsys=endpoint total=1m33.651333003s waitingForCTClean=1m33.646810669s waitingForLock="1.85µs"
```

A few questions: 
1) Could a long CT cleaning cycle cause endpoint (re)generation to fail? We are seeing 500s in PUTs and we're not sure if this is a cause or a symptom.
2) Are there parameters we can tune to reduce this time such as reducing conntrack-gc-interval?
3) Why is k8sPodName blank?

Thanks in advance :)

**Comment by joestringer** [comment]
Hi @jeff-an,
(1) That's not really expected.. do you have some indicators that suggest it's CT GC that is running during that long cycle? Perhaps you could use gops to inspect Cilium cpu usage during that long period.
(2) There's a few *conntrack* settings, you're welcome to tinker with them but I'm not really sure whether this will impact the problem  you're facing.
(3) That sounds like a separate bug. I filed #22242 for that one.

**Comment by jeff-an** [comment]
Thanks for the response!

The only evidence is the "waitingForCTClean=1m33.646810669s" attribute on the first log I sent. We do know CPU can be spikey on these machines, so it is possible Cilium is getting thrashed. When we have more headroom the cluster, we don't see this issue. What's interesting though is that even when CPU has come down on the node, new requests continue to result in 429s (though this could be a separate issue relating to the dynamic rate limiter kicking in).

**Comment by julianwiedmann** [comment]
> Regarding the "too many syscalls" issue, I've been informed that as of Linux v5.6 there is batched map lookup which should allow us to make this more efficient on newer kernels:
> 
> https://lore.kernel.org/bpf/20191119014357.98465-3-brianvv@google.com/

I believe this has been addressed now by @tommyp1ckles in https://github.com/cilium/cilium/pull/36288.

**Comment by tommyp1ckles** [comment]
@julianwiedmann @joestringer Yup #36288 adds using batched iterators (when supported).  Benchmarking suggests this will decrease the amount of time spent doing iterations by ~5-8x.

This is primarily for the iteration portion (i.e. the benchmark was testing a "sparse" scenario where we only have to iterate the map but find few elements marked for GC),  performance would once again degrade significantly if we need to do a large number of deletes.

As a followup, we can use the `bpf_batch_delete` command to batch up deletes before submitting them. The one caveat is that the bpf_batch_delete stops deleting and immediately returns `ENOENT` if it encounters a map element that is not in the map (something that could easily happen with the CT LRU map).  It does write back a number of deleted elements to userspace which could be used to restart deleting at the n+1th element but in the worst case would behave no better than the individual `bpf_delete_element` calls.

I'm planning to submit a kernel patch to add a way to do unequivocal batched deletes in the coming days, but assuming that would be accepted - would be a while before we could meaningfully make use of that.

**Comment by arighnac** [comment]
Hi @joestringer @julianwiedmann @tommyp1ckles We have been internally grappling with the problem of high endpoint create/delete latency, and were exploring bpf iterators to perform an efficient lookup of the conntrack table to garbage collect all entries for the endpoint to be created/deleted. Filtering of the IP can then happen within the kernel space, reducing syscall overhead and effectively reducing overall endpoint creation latency, magnified with high pod churn.
Curious to know if we have ever thought about using BPF iterators to optimize for CTmap GC..
